week4/wordbreak.cc

```diff
@@ -0,0 +1,136 @@
+/* wordbreak.cc
+ * author: connie
+ *
+ * WORD BREAK
+ * Given an input string and a dictionary of words,
+ * find out if the input string can be segmented into a
+ * space-separated sequence of dictionary words.
+ *
+ */
+#include <iostream>
+#include <sstream>
+#include <cassert>
+#include <vector>
+#include <string>
+#include <map>
+
+using namespace std;
+
+//== DECLARATIONS
+bool wordInDict(string word, vector<string>& dict);
+bool wordBreak_RC(string s, vector<string>& dict);
+bool wordBreak_DP(string s, vector<string>& dict); // TODO w/ map
+
+//== DEFINITIONS
+//#define DEBUG
+
+//== FUNCTIONS
+/*
+ * bool wordInDict(string word, vector<string>& dict)
+ * inputs: word - word to check
+           dict - the dictionary (array of strings) with valid words
+ * outputs: True for valid dictionary word, False for invalid word
+ * notes: helper function for iterating thru the array
+ */
+bool wordInDict(string word, vector<string>& dict)
+{
+  if (word.size() == 0) return false;
+
+  // iterate thru dictionary to check if the word is valid
+  for (string dictword : dict) {
+#ifdef DEBUG
+    cout << word << " ?= " << dictword << endl;
+#endif
+    if (word.compare(dictword) == 0) return true;
+  }
+
+  return false;
+}
+
+/*
+ * bool wordBreak_RC(string s, vector<string>& dict)
+ * inputs: s - the string to check for a valid "word break"
+           dict - the dictionary (array of strings) with valid words
+ * outputs: True for valid word break, False for invalid word break
+ * notes: the recursive implementation that verifies progressive substrings
+ */
+bool wordBreak_RC(string s, vector<string>& dict)
```

Comment:
Very nice job with the brute force approach :) I saw your comment and you're right, dynamic programming is a good way to optimize the solution to this problem